stl/inc/xutility

```diff
@@ -5087,75 +5108,149 @@ _NODISCARD constexpr auto lexicographical_compare_three_way(
 }
 #endif // __cpp_lib_concepts
 
-template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, true_type, true_type, _Any_tag, false_type) {
-    // signed _Elem, signed _Ty
-    return SCHAR_MIN <= _Val && _Val <= SCHAR_MAX;
-}
-
-template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, true_type, false_type, true_type, false_type) {
-    // signed _Elem, unsigned _Ty, -1 == static_cast<_Ty>(-1)
-    return _Val <= SCHAR_MAX || static_cast<_Ty>(SCHAR_MIN) <= _Val;
-}
-
-template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, true_type, false_type, false_type, false_type) {
-    // signed _Elem, unsigned _Ty, -1 != static_cast<_Ty>(-1)
-    return _Val <= SCHAR_MAX;
-}
-
-template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, false_type, true_type, _Any_tag, false_type) {
-    // unsigned _Elem, signed _Ty
-    return 0 <= _Val && _Val <= UCHAR_MAX;
-}
-
-template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, false_type, false_type, _Any_tag, false_type) {
-    // unsigned _Elem, unsigned _Ty
-    return _Val <= UCHAR_MAX;
-}
-
-template <class _Ty>
-_NODISCARD constexpr bool _Within_limits(const _Ty& _Val, _Any_tag, _Any_tag, _Any_tag, true_type) {
-    // bool _Elem
-    return _Val == true || _Val == false;
-}
-
 template <class _InIt, class _Ty>
 _NODISCARD constexpr bool _Within_limits(const _InIt&, const _Ty& _Val) {
     // check whether _Val is within the limits of _Elem
-    using _Elem = _Iter_value_t<_InIt>;
-    return _Within_limits(_Val, bool_constant<is_signed_v<_Elem>>{}, bool_constant<is_signed_v<_Ty>>{},
-        bool_constant<-1 == static_cast<_Ty>(-1)>{}, bool_constant<is_same_v<_Elem, bool>>{});
-}
-
-template <class _InIt>
-_NODISCARD constexpr bool _Within_limits(const _InIt&, const bool&) { // bools are always within the limits of _Elem
-    return true;
-}
-
+    if constexpr (is_same_v<_Ty, bool>) {
+        return true;
+    }
 #ifdef __cpp_lib_byte
-template <class _InIt>
-_NODISCARD constexpr bool _Within_limits(const _InIt&, const byte&) { // bytes are only comparable with other bytes
-    return true;
-}
+    else if constexpr (is_same_v<_Ty, byte>) {
+        return true;
+    }
 #endif // __cpp_lib_byte
+    else {
+        using _Elem = _Iter_value_t<_InIt>;
+        _STL_INTERNAL_STATIC_ASSERT(
+            sizeof(_Elem) == 1 || sizeof(_Elem) == 2 || sizeof(_Elem) == 4 || sizeof(_Elem) == 8);
+        if constexpr (is_same_v<_Elem, bool>) {
+            return _Val == true || _Val == false;
+        } else if constexpr (is_signed_v<_Elem>) {
+            if constexpr (is_signed_v<_Ty>) {
+                // signed _Elem, signed _Ty
+                if constexpr (sizeof(_Elem) == 1) {
+                    return SCHAR_MIN <= _Val && _Val <= SCHAR_MAX;

```

Comment:
To avoid braching on size here, compute limits using two's complement, shift and `CHAR_BIT`